### PR TITLE
Rules repository for Application ID Detection

### PR DESCRIPTION
**Pull request by Demair**
Rule repository for application ID detection based on the OpenAppID preprocessor has been added.

### CODE REVIEW

**Comment by pfSense-git-updates** [comment]
Before this pull request can be accepted you must first sign a CLA as described at https://www.pfsense.org/about-pfsense/#cla. Please read for more details.

File: security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_defs.inc [hunk 1]
```diff
@@ -48,11 +48,13 @@ if (!defined("SNORT_IPREP_PATH"))
...
 if (!defined('SNORT_APPID_ODP_PATH'))
 	define('SNORT_APPID_ODP_PATH', SNORTDIR . "/appid/");
+if (!defined('SNORT_APPID_RULES_PATH'))
+        define('SNORT_APPID_RULES_PATH', SNORTDIR . "/preproc_rules/");
 
 // Rules filenames, download URLs and prefixes.
 // Be sure to include the trailing backslash on URLs.
 if (!defined("SNORT_ENFORCING_RULES_FILENAME"))
-	define("SNORT_ENFORCING_RULES_FILENAME", "snort.rules");
+        define("SNORT_ENFORCING_RULES_FILENAME", "snort.rules");
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_defs.inc]: Please don't break indent, restore TAB here

File: security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_generate_conf.php [hunk 2]
```diff
@@ -2,23 +2,33 @@
...
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_generate_conf.php]: Restore license and copyright here

File: security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_migrate_config.php [hunk 3]
```diff
@@ -2,22 +2,29 @@
...
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_migrate_config.php]: Restore license and copyright here

File: security/pfSense-pkg-snort/files/usr/local/www/snort/snort_download_updates.php [hunk 4]
```diff
@@ -1,23 +1,38 @@
...
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/www/snort/snort_download_updates.php]: Restore license and copyright

File: security/pfSense-pkg-snort/files/usr/local/www/snort/snort_rules.php [hunk 5]
```diff
@@ -171,6 +171,10 @@ function add_title_attribute($tag, $title) {
 $categories[] = "decoder.rules";
 $categories[] = "preprocessor.rules";
 $categories[] = "sensitive-data.rules";
+if (file_exists("{$snortdir}/snort_{$snort_uuid}_{$if_real}/preproc_rules/appid.rules"))
+{
+$categories[] = "appid.rules";
+}
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/www/snort/snort_rules.php]: Fix indent

**Comment by rbgarga** [comment]
@bmeeks8 I would like to hear from you about this change

**Comment by bmeeks8** [comment]
I like the concept of the changes.  They provide a set of ready-to-use rules for the Open AppID functionality Cisco added to Snort a couple of years ago.  I would like to see the option implemented a bit differently though.  Right now the new rules are an "all or none" proposition in that they are either all enabled or all disabled.  There is no way within the GUI for the admin to customize the rules by turning some off and leaving others on like can be done for all the other Snort rule sets.  For example, there is no display of these Open AppID rules on the RULES tab.  The RULES tab is where individual rules can be turned on or off.

So I would like to see the AppID rules in this change treated the same as all the other rules files so that individual AppID rules can be toggled on or off instead of the current situation where you either enable all of the AppID rules or none of them.  This would mean incorporating these new AppID rules into the RULES tab.

I am also a bit concerned with hard-coding the URL and filename for these rules unless the source is guaranteed to be reliable.  In other words, the source appears to be a university in Brazil.  Will these rules be maintained there and always be available via the hard-coded URL?  Would it perhaps be better to allow the pfSense firewall admin the ability to customize or edit the URL and filename within the GUI (for example, on the GLOBAL tab)?

Bill

**Comment by bmeeks8** [comment]
I have one other suggestion for the submitter.  Can the large single file of AppID rules in the gzip archive be split out into smaller files and the contents grouped by rule type?  For example, have an AppID rules file for social media apps; another AppID rules file for games, etc.  In this way the AppID rules can be presented in a format similar to the other Snort rules where you have server rules, web client rules, DNS rules, etc., all in separate rules files.

**Comment by Demair** [comment]
Good morning Bill Meeks, defaults rules in the appid.rules file are deprecated #rules, the admin has to go to Category Selection: and activates the rules, after doing so the rule will be loaded into the snort.rules file only the rule that the Admin pfsense enabled.

**Comment by bmeeks8** [comment]
And one final comment -- I downloaded the gzip archive and took a peek at the AppID rules within.  The single archive is quite large (390K), so that is one good reason to split the contents into a series of smaller files with their content organized by application type.  I also noticed that in the current file **all** of the rules are commented out.  This means as currently provided none of the AppID rules would actually be enforced.  What mechanism would the admin use, in the current code, to enable rules in this archive?  And if manual editing of the file is required, would not all edits be overwritten on the next automatic update?

**Comment by Demair** [comment]
Good morning Bill Meeks, defaults rules in the appid.rules file are deprecated #rules, the admin has to go to Category Selection: and activates the rules, after doing so the rule will be loaded into the snort.rules file only the rule that the Admin pfsense enabled.

**Comment by bmeeks8** [comment]
@Demair:  I do not see where in the code changes submitted the comment symbol "#" at the front of each rule in the archive is removed.  All I see is a place where the AppID rules file is added to the array of rules to load.  Simply loading a file is not everything.  Within a given file, if a rule is commented out, it will not be scanned and enforced.

**Comment by Demair** [comment]
The url is guaranteed the board of the university has total interest in keeping both the url and the rules, always updated and working. The idea of splitting the file in part is interesting.

**Comment by Demair** [comment]
The idea that the rules of appid.rules work the same as the mechanism of sensitive-data.rules, in the appid.rules file all the rules are commented out, and will be activated via the file snort.rules, but we can do as you think best done of this It was the easiest way for me.

**Comment by bmeeks8** [comment]
OK, I did miss the part where they are shown on the RULES tab.  Scanned the changes too quickly ... <grin>.

I do still recommend perhaps splitting the new rules into groups of similar rules and put each group in its own AppID file.  You could still group all the files into a single gzip archive just like the other Snort rules.  Having the AppID rules organized by application type would be helpful.

**Comment by Demair** [comment]
OK, the idea of splitting the rules by partner media applications, DNS rules, etc., is great, I'll be separated into smaller files and with the pattern you mentioned.

**Comment by Demair** [comment]
Good morning, I finished dividing the files into categories today and I still commit. Was created a category network_protocol if the user ticks without checking can block access to the main services like http, dns among others. What I will leave the most usable protocols with the #, any suggestions ????

**Comment by rbgarga** [comment]
@Demair licenses are still broken. Now you have a mix of BSD and Apache. It should look like exactly like this example:

https://github.com/pfsense/FreeBSD-ports/blob/devel/devel/php56-pfSense-module/files/pfSense.c#L2

**Comment by Demair** [comment]
@rbgarga Good night, my mistake added the difference in the files and I did not check. excuse me

File: security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_generate_conf.php [hunk 6]
```diff
@@ -17,8 +18,9 @@
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
+ * See the License for the specific language governing permissions and 
  * limitations under the License.
+ *
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_generate_conf.php]: All above changes to license text are unnecessary and should be reverted

File: security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_migrate_config.php [hunk 7]
```diff
@@ -18,6 +19,7 @@
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
+ *
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_migrate_config.php]: All above changes to license text are unnecessary and should be reverted

File: security/pfSense-pkg-snort/files/usr/local/www/snort/snort_download_updates.php [hunk 8]
```diff
@@ -13,6 +16,7 @@
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
+ *
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/www/snort/snort_download_updates.php]: All above changes to license text are unnecessary and should be reverted

File: security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_generate_conf.php [hunk 9]
```diff
@@ -2,10 +2,10 @@
 /*
  * snort_generate_conf.php
  *
- * part of pfSense (https://www.pfsense.org)
+ * part of pfSense (https://www.pfsense.org
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_generate_conf.php]: Unnecessary change on license, revert

File: security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_generate_conf.php [hunk 10]
```diff
@@ -2,10 +2,10 @@
...
  * Copyright (c) 2006-2016 Rubicon Communications, LLC (Netgate)
  * Copyright (c) 2009-2010 Robert Zelaya
- * Copyright (c) 2013-2014 Bill Meeks
+ * Copyright (C) 2013,2014 Bill Meeks
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_generate_conf.php]: Unnecessary change on license, revert

File: security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_generate_conf.php [hunk 11]
```diff
@@ -17,7 +17,7 @@
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
+ * See the License for the specific language governing permissions and
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_generate_conf.php]: Unnecessary change on license, revert

File: security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_migrate_config.php [hunk 12]
```diff
@@ -3,8 +3,8 @@
  * snort_migrate_config.php
  *
  * part of pfSense (https://www.pfsense.org)
- * Copyright (c) 2016 Rubicon Communications, LLC (Netgate)
- * Copyright (c) 2013-2014 Bill Meeks
+ * Copyright (c) 2006-2016 Rubicon Communications, LLC (Netgate)
+ * Copyright (C) 2013, 2014 Bill Meeks
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/pkg/snort/snort_migrate_config.php]: Unnecessary change on license, revert

File: security/pfSense-pkg-snort/files/usr/local/www/snort/snort_download_updates.php [hunk 13]
```diff
@@ -4,7 +4,7 @@
  *
  * part of pfSense (https://www.pfsense.org)
  * Copyright (c) 2004-2016 Rubicon Communications, LLC (Netgate)
- * Copyright (c) 2016 Bill Meeks
+ * Copyright (C) 2016 Bill Meeks
```
Comment (rbgarga) [inline security/pfSense-pkg-snort/files/usr/local/www/snort/snort_download_updates.php]: Unnecessary change on license, revert

**Comment by rbgarga** [comment]
Lets wait for @bmeeks8 related to the categories division before go ahead

**Comment by Demair** [comment]
@rbgarga  @bmeeks8 For information purposes I used the base of http://appwiki.checkpoint.com/appwikisdb/public.htm and https://applabs.proceranetworks.com/aks/applications to catalog the applications.

**Comment by rbgarga** [comment]
@bmeeks8 did you have a chance to check last version?

**Comment by bmeeks8** [comment]
I'm sorry.  I have been remiss in looking at these changes on GitHub.  I am in agreement with them and OK for the changes to be accepted.  I really appreciate the work @Demair has put into improving the Snort package.  This should make the OpenAppID functionality that has been in the package for a while much easier for folks to actually use since they will not have to create all of the required rules themselves.  They now can choose from a list of ready-made rules sorted by category.

Bill
